src/Compiler.Dynamic/SystemWebCompilation.cs

```diff
@@ -34,19 +35,56 @@ internal abstract class SystemWebCompilation : IPageCompiler, IDisposable
         PagesSection.Instance = pagesSection.Value;
     }
 
-    public Task<ICompiledPage> CompilePageAsync(IFileProvider files, string path, CancellationToken token)
+    public async Task<ICompiledPage> CompilePageAsync(IFileProvider files, string path, CancellationToken token)
     {
         try
         {
-            return Task.FromResult(CompilePage(files, path, token));
+            if (_compiled.TryGetValue(path, out var existing))
```

Comment:
make sense to pick up/reuse and make it available to assembly rather re-compiling..